network/connection.go

```diff
@@ -0,0 +1,379 @@
+		process: proc,
+		Started: timestamp,
+	}
+}
+
+// GetConnection fetches a Connection from the database.
+func GetConnection(id string) (*Connection, bool) {
+	connsLock.RLock()
+	defer connsLock.RUnlock()
+
+	conn, ok := conns[id]
+	return conn, ok
+}
+
+// Accept accepts the connection.
+func (conn *Connection) Accept(reason string) {
+	if conn.SetVerdict(VerdictAccept) {
+		conn.Reason = reason
+		log.Infof("filter: granting connection %s, %s", conn, conn.Reason)
+	} else {
+		log.Warningf("filter: tried to accept %s, but current verdict is %s", conn, conn.Verdict)
+	}
+}
+
+// Block blocks the connection.
+func (conn *Connection) Block(reason string) {
+	if conn.SetVerdict(VerdictBlock) {
+		conn.Reason = reason
+		log.Infof("filter: blocking connection %s, %s", conn, conn.Reason)
+	} else {
+		log.Warningf("filter: tried to block %s, but current verdict is %s", conn, conn.Verdict)
+	}
+}
+
+// Drop drops the connection.
+func (conn *Connection) Drop(reason string) {
+	if conn.SetVerdict(VerdictDrop) {
+		conn.Reason = reason
+		log.Infof("filter: dropping connection %s, %s", conn, conn.Reason)
+	} else {
+		log.Warningf("filter: tried to drop %s, but current verdict is %s", conn, conn.Verdict)
+	}
+}
+
+// Deny blocks or drops the link depending on the connection direction.
+func (conn *Connection) Deny(reason string) {
+	if conn.Inbound {
+		conn.Drop(reason)
+	} else {
+		conn.Block(reason)
+	}
+}
+
+// SetVerdict sets a new verdict for the connection, making sure it does not interfere with previous verdicts.
+func (conn *Connection) SetVerdict(newVerdict Verdict) (ok bool) {
+	if newVerdict >= conn.Verdict {
+		conn.Verdict = newVerdict
+		return true
+	}
+	return false
+}
+
+// Process returns the connection's process.
+func (conn *Connection) Process() *process.Process {
+	return conn.process
+}
+
+// SaveWhenFinished marks the connection for saving it after the firewall handler.
+func (conn *Connection) SaveWhenFinished() {
+	conn.saveWhenFinished = true
+}
+
+// save saves the link object in the storage and propagates the change.
+func (conn *Connection) save() {
+	if conn.ID == "" {
+
+		// dns request
+		if !conn.KeyIsSet() {
+			conn.SetKey(fmt.Sprintf("network:tree/%d/%s", conn.process.Pid, conn.Scope))
+			conn.UpdateMeta()
+		}
+		// save to internal state
+		// check if it already exists
+		mapKey := strconv.Itoa(conn.process.Pid) + "/" + conn.Scope
+		dnsConnsLock.RLock()
+		_, ok := dnsConns[mapKey]
+		dnsConnsLock.RUnlock()
+		if !ok {
+			dnsConnsLock.Lock()
+			dnsConns[mapKey] = conn
+			dnsConnsLock.Unlock()
+		}
+
+	} else {
+
+		// connection
+		if !conn.KeyIsSet() {
+			conn.SetKey(fmt.Sprintf("network:tree/%d/%s/%s", conn.process.Pid, conn.Scope, conn.ID))
+			conn.UpdateMeta()
+		}
+		// save to internal state
+		// check if it already exists
+		connsLock.RLock()
+		_, ok := conns[conn.ID]
+		connsLock.RUnlock()
+		if !ok {
+			connsLock.Lock()
+			conns[conn.ID] = conn
+			connsLock.Unlock()
+		}
```

Comment:
Same as above, I'd just Lock() and drop the RLock() instead of trying to get the lock twice

Reply:
See comment above.